### ISSUE TITLE
[question] APNs integration

### ISSUE THREAD

**Issue by vonovak** [notifee/react-native-notifee#90]
Hello!
I've got some questions regarding integrating with APNs because I'm not sure where "notifee's responsibility" starts and ends.

The idea is following: on iOS, I'm not using firebase and so I'd like to use https://github.com/react-native-community/push-notification-ios to _get APNs token only_ (and deliver it to a server). That means I only implement

```
- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
  [RNCPushNotificationIOS didRegisterForRemoteNotificationsWithDeviceToken:deviceToken];
}

- (void)application:(UIApplication *)application didFailToRegisterForRemoteNotificationsWithError:(NSError *)error {
  [RNCPushNotificationIOS didFailToRegisterForRemoteNotificationsWithError:error];
}
```

and then I want to use Notifee to react to when notification is received (both when app is in background and foreground) and when it is pressed by user, including actions offered in the notification. It appears that [events](https://notifee.app/react-native/docs/events) should handle just that - ?

So far I was able to get notification delivered on android with fcm, and the `onForegroundEvent` as well as `onBackgroundEvent` are sent over to the app (I just made a console.log there and found it in logcat).

When I tried to do the same on ios, I only managed to get the push notification delivered when the app is in background / killed (so remote notifications are delivered, getting the APNs token seems to work). Now, when I try to subscribe for the events, I unfortunately don't receive any :/ 

I'm aware of this [comment](https://github.com/notifee/react-native-notifee/issues/1#issuecomment-625808121) and the [implementation](https://github.com/notifee/react-native-notifee/blob/2da633ae49fefdc7a54a0a09b60b27b75495928e/src/NotifeeApiModule.ts#L233). 

Unfortunately, since the native code is not open, I'm not quite sure what is going on inside and I'm not sure if my expectations are correct.

in device logs, I found

```
Warning: UNUserNotificationCenter delegate received call to -userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler: but the completion handler was never called.
```

which I'm guessing might be an issue?

I'd like to ask if my expectations are correct and if my handlers (at least the foreground one) should be called. I'm registering the handlers early in the app, is it theoretically possible that the event is fired before a listener is attached? (I would understand that for the background event but not for the foreground one?). Lastly, can someone confirm that this works for them using RN 60? Maybe I should upgrade first...

**Comment by Salakar** [comment]
Some quick answers;

> (2) is it theoretically possible that the event is fired from native before a listener is attached in JS? (I would understand that for the background event but not for the foreground one?). Does notifee's native code somehow wait until JS "is ready"?

On RNFirebase we have a temporary buffer of all events that are sent while JS is not ready that is uncorked once JS is ready, this is something we need to apply to Notifee still though in our Notifee testing we could never get events to fire before JS was ready


> (3) does notifee implement willPresentNotification and didReceiveNotificationResponse for its events functionality?

Yes, but only handles notifications that were created explicitly by Notifee, Notifees handlers for willPresentNotification and didReceiveNotificationResponse ignore everything else, leaving things like RNFirebase messaging to handle it. Though if you're having issues here then I suspect we may need to do some more work here 🤔 

> (4) any ETA for proper onBackgroundEvent support on iOS?

Hoping to look at it this week.

> (5) how is onBackgroundEvent different from RNFB's messaging().setBackgroundMessageHandler? Maybe I'm mixing local and remote notifications and asking stuff that should be obvious but I didn't feel like I found answers for these in the docs. 

Notifee `onBackgroundEvent` and RNFB `setBackgroundMessageHandler` are pretty much identical in behaviour - however Notifee is local notifications only, so only events from notifications created via Notifee apis will come through to it. Whereas RNFB is FCM (remote) messages only and will ignore Notifee events as they're not FCM.

> Generally, when reading docs, I keep noticing it's always documented what some library is but it's often not documented what the library is not.

This is a good point, cc @Ehesp - we could be clearer on what Notifee is not, e.g. not for remote notifications, not a FCM replacement, etc.

>  But otherwise the docs are really nice (I really like click-though type definitions)!

Thank you!


----

If you want to go through anything in more depth though and debug any issues, feel free to ping me on Discord 👍

**Comment by vonovak** [comment]
> Yes, but it only handles notifications that were created explicitly by Notifee, Notifee's handlers for willPresentNotification and didReceiveNotificationResponse ignore everything else, leaving things like RNFirebase messaging to handle it.

okay, this makes it clear now. I wanted to get events fo remote notifications which is not supported.

Any plan to allow responding to remote notifications too, or do you to purposely avoid it and leave in the scope of other libraries? I guess I'll try my luck with [react-native-notifications](https://github.com/wix/react-native-notifications), it just feels funny I have to add another dep because it seems Notifee could get the job done too.

Thanks!

**Comment by Salakar** [comment]
Its probably something we should add otherwise you'd need yet another library, I think the initial issue was that if it didn't go through Notifee then we only had access to a limited amount of data from the notification and many fields could not be read/populated for sending the event - though in this scenario I think as long as you have notification data & action ids that should cover everything.

Will look into it

**Comment by jyliang** [comment]
@Salakar could you please let me know if Notifee becomes the delegate for UNUserNotificationCenter .currentNotificationCenter() ? If so, it looks like some of the events might be stopping inside Notifee.

For example I get `didReceiveNotificationResponse` called inside `RNFBMessaging+UNUserNotificationCenter.m', it appears the 'originalDelegate' is `NotifeeCoreUNUserNotificationCenter`. This is a local notification I scheduled on the native side, so I don't expect Notifee to pick it up. However my original delegate is not called. I can open up another ticket if needed

**Comment by vonovak** [comment]
@jyliang yes, from my experience, notifee overwrites the delegate you may have set previously, and so your delegate does not learn the events it should. I hope this will be addressed at some point soon.

**Comment by helenaford** [comment]
Notifee now calls the original delegate in the latest release v0.11.1.

**Comment by vonovak** [comment]
hello!
I'd like to ask if there's any ETA on exposing `didReceiveNotificationResponse` (support for JS events that correspond to responses to remote notifications) to JS? We have discussed this with @Salakar on discord already a few months back and it seemed like the support for this was going to be added soon (Mike said it'd be easy to add) - should I count on this or should I somehow work around this? I'm available to add the feature to notifee, too.

Thanks!

**Comment by vonovak** [comment]
@Salakar @helenaford hello and sorry for pinging again, could you please provide an answer? Thank you!

**Comment by helenaford** [comment]
Hey, @vonovak sorry for the delayed response.  We are talking it over internally and can back to you at the end of this week. From initial talks, I think the answer is `yes`, we can implement it. 

Also, from reading your initial comment on this thread, do you still need to know when a remote notification is delivered? Event could be something like `REMOTE_NOTIFICATION_DELIVERED'.

And, do you only need iOS support?

**Comment by vonovak** [comment]
@helenaford thanks for the reply. I thought about this again for a bit; I know you folks probably don't want to get into handling too much "remote notifications" stuff.


At the same time, as a consumer of the library and by looking at the docs (https://notifee.app/react-native/docs/events#foreground-events) I had the impression that if I tap one of the actions in the notification, then notifee will handle it. When I respond to notification action on android, the `notifee.onForegroundEvent` as well as `notifee.onBackgroundEvent` do get called, however it was not the case on iOS. So in this case, it was the matter of having consistent behavior on both platforms - that it the main issue I had with current state of things, so think about it and you'll see.

> do you still need to know when a remote notification is delivered?

actually, I do not. At least not now 😆  I'm "juggling" several things here: data-only notifications on android, but "normal notification" (no content-available), react native firebase on android, but not on iOS (just using apns directly because they don't want to go through any extra layers of abstraction), so it's a bit of a complicated story. But given that 

> Notifee now calls the original delegate in the latest release v0.11.1.

I guess I'll be able to work out potential issues somehow by myself.

Thank you!

**Comment by vonovak** [comment]
ah, I'm sorry, I just realized - because it's a data-only notification on Android, I use `notifee.displayNotification` to present it. And so when user responds to the notification, the notifee handler is called. But on iOS, we use plain notification, which is why notifee ignores it (it was not displayed by notifee). It makes sense to me now, sorry for my ignorance. Maybe a little bit of clarification in the docs would be nice.

**Comment by mikehardy** [comment]
@vonovak where in the documentation would you put it? I think on the handling events - right here? https://notifee.app/react-native/docs/events#listening-for-events

Also, it is perhaps surprising but then understandable that the library only emits events for notifications it posts, but...I wonder if there is room philosophically / ability technically to mark a custom generated notification as being "Notifee Event Capable" such that even if Notifee did not generate the notification, it would react to it. That might tie everything up for you? I'm not saying it is possible, this more of a 🤔  question to see if it would be useful even

**Comment by helenaford** [comment]
@vonovak no problem, just having this conversation is helpful to see what can be improved... 

Think we're on the same lines @mikehardy. Could be possible to put something in the `apns` payload to let `notifee` know they can handle the notification and send events. I think it's simple on iOS but the challenge is to make sure both platforms are consistent. So, if we do support these extra events on iOS, we'd need to make sure Android does too.

**Comment by vonovak** [comment]
> where in the documentation would you put it? I think on the handling events - right here? https://notifee.app/react-native/docs/events#listening-for-events

yes, that sounds like the best place


> I wonder if there is room philosophically / ability technically to mark a custom generated notification as being "Notifee Event Capable" such that even if Notifee did not generate the notification, it would react to it. That might tie everything up for you?

I think I just approached the library with wrong expectations from the start. Eg. if I look at [userNotificationCenter:didReceiveNotificationResponse:withCompletionHandler:](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649501-usernotificationcenter?language=objc) I thought I would receive the user's response in one of the notifee's event handlers regardless of whether the notification was a remote one or one from notifee, because I thought "why would a library differentiate between them?". And if it only handles responses only to its own notifications, how am I supposed to handle the other ones? Should I need to use another library for that? Why? Just explaining what my initial expectations were, and somehow the docs didn't correct that mental model :)
